### PR TITLE
[url] skip parsing username:password@ for udp/rtp streams

### LINKED ISSUE #17220: RTP streams that use an interface IP incorrectly interpreted as username
### Describe the bug
Here is a clear and concise description of what the problem is:

As an example if a stream has this URL: `rtp://87.141.215.251@232.0.20.35:10000`

The `87.141.215.251` will be interpreted as the username instead of being passed at part of the URL.

Reported in forum post: https://forum.kodi.tv/showthread.php?tid=350901

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

The `87.141.215.251` will be interpreted as part of the URL.

## Actual Behavior

The `87.141.215.251` will be interpreted as A username.

## Possible Fix

Check when extracting a username if it’s an IP. If not then leave the URL as is. Not sure of the impact/side effects of doing this.

### To Reproduce

1. Add an RTP stream with the above format to IPTV Simple 
2. Play the stream/channel
3. Stream not recognised error

### Debuglog

The debuglog can be found here:

From forum post: https://paste.kodi.tv/wufogamapi.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [X] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Ubuntu 16.04 LTS
 - Kodi version: 18.5

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

Username and password are not valid for UDP/RTP streams. This can cause issues where the URL is in the form rtp://sourceip@multicastip.

Fixes: https://github.com/xbmc/xbmc/issues/17220

## Motivation and Context

Fix udp/rtp streams in the format rtp://sourceip@multicastip

## How Has This Been Tested?

Requires user testing. I don't have a stream of this type.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

**Comment by phunkyfish** [comment]
Fixed @Rechi

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

**Comment by phunkyfish** [comment]
It's really fixed now.

**Comment by phunkyfish** [comment]
After some user testing we get an error with what is enabled in the ffmpeg build:

```
20-01-20 20:30:40.851 T:41920  NOTICE: Creating Demuxer
2020-01-20 20:30:40.851 T:41920   DEBUG: ffmpeg[0x281d99ca380X]: [udp] 'circular_buffer_size' option was set but it is not supported on this build (pthread support is required)
2020-01-20 20:30:40.853 T:23612   DEBUG: Previous line repeats 1 times.
2020-01-20 20:30:40.853 T:23612   DEBUG: Loading settings for pvr://channels/tv/All%20channels/pvr.iptvsimple_425868680.pvr
```

Full log: https://paste.kodi.tv/vewologahu.kodi

**Comment by phunkyfish** [comment]
As it’s lacking pthread the error should be windows only.

**Comment by phunkyfish** [comment]
@Paxxi would there be any side effects if we build ffmpeg for windows using `pthreads` instead `w32threads` that you know of?

Sp with something like `--disable-w32threads --enable-pthreads` or it's working equivalent?

**Comment by Paxxi** [comment]
I don't really know.

Based on the given information I would assume pthreads don't work on windows or there wouldn't really be a reason to have a w32-threads option.

Maybe it's a simple patch for ffmpeg where it checks for thread support? I would look into that first, if it's more than a bad if check then it might be worth trying with pthreads.

**Comment by phunkyfish** [comment]
Good point. Will look into it.

**Comment by phunkyfish** [comment]
Ok, so ffmpeg requires the URL in a different format if source IPs are used: http://ffmpeg.org/ffmpeg-protocols.html#rtp

So ` rtp://sourceip@multicastip` needs to become ` rtp://@multicastip?sources=sourceip`. Which will mean an additional change in DVDDemuxFFmpeg so support this.

**Comment by phunkyfish** [comment]
Also, after some reading there is a marginal difference between w32threads and pthreads, hence why most use w32. We can always try it later on if we eventually need to.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 1]
```diff
@@ -828,6 +850,28 @@ AVDictionary* CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput()
...
+      {
+        // sourceip found
+        std::string strSourceIp = strURL.substr(start, found - start);
+        CLog::Log(LOGDEBUG,
+                  "CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput() UDP/RTP sources '%s'",
+                  strSourceIp.c_str());
+        av_dict_set(&options, "sources", strSourceIp.c_str(), 0);
+        av_dict_set(&options, "connect", "1", 0);
+        av_dict_set(&options, "write_to_source", "1", 0);
+        //av_dict_set(&options, "buffer_size", "200000", 0);
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: If this is not needed, please remove the line.

Comment (phunkyfish) [reply xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: Will do @Rechi, as the original fix did not work entirely I've added a WIP tag.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 2]
```diff
@@ -264,6 +264,38 @@ bool CDVDDemuxFFmpeg::Open(std::shared_ptr<CDVDInputStream> pInput, bool streami
...
+        found = strURL.find("@");
+
+        if (found != std::string::npos && found > start)
+        {
+          // sourceip found
+          std::string strSourceIp = strURL.substr(start, found - start);
+
+          strFile = strURL.substr(0, start);
+          strFile += strURL.substr(found);
+          if(strFile.back() == '/')
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: > Control statement keywords have to be separated from opening parentheses by one space.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

**Comment by phunkyfish** [comment]
@Paxxi so it looks like in order to get this to work on windows pthreads will be required. We did it in ffmpegdirect like this building it as part of addon depends: https://github.com/xbmc/inputstream.ffmpegdirect/pull/10/commits/df47c9f61db083362cb9b91aa056bfbb77a4abba

How would you recommend to go about this so we can try including pthreads here?: https://github.com/xbmc/xbmc/blob/master/tools/buildsteps/windows/buildffmpeg.sh

**Comment by Paxxi** [comment]
What part of ffmpeg is it that you're using?

**Comment by phunkyfish** [comment]
This would be libavformat in order to demux the stream.

**Comment by FernetMenta** [comment]
I seriously doubt that changing ffmpeg to pthreads is an appropriate solution for the issue you are trying to solve. Try to explain why this should require pthreads. Makes no sense at all. Why don't you elaborate on the root cause. Have you asked ffmpeg developers already?

Changing ffmpeg to pthreads introduces a high risk of breakage. Taking this risk for something that should be fixed elsewhere is a bad idea.

**Comment by phunkyfish** [comment]
UDP streams use pthreads for their  circular_buffer logic. The included patch adds missing logic in w32pthreads to allow this (hopefully) to function, but just for UDP.

Have not asked ffmpeg devs yet. I have submitted a patch that fixes an RTP bug, but I fear this currently lacks a maintainer.

Is the ffmpeg patch in the PR not how you would go about this?

**Comment by FernetMenta** [comment]
Your commit messages are misleading. I thought you were disabling w32threads in favour of pthreads for the mingw build. This would have been a major change. 

What you actually do is making udp use the pthreads compat layer on windows platform. That's a completely different story. Sorry for thy noise. More precise commit messages welcome :)

**Comment by phunkyfish** [comment]
:) in hindsight it is a bit ambiguous!

**Comment by phunkyfish** [comment]
All user testing on this has passed so will merge this tomorrow. 

Due to the patching required I don’t think I will backport this to Leia.

**Pull request merged**